src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java

```diff
@@ -250,109 +307,201 @@ private UserDetails retrieveUser(String username, UsernamePasswordAuthentication
-                    closeQuietly(test);
-                }
-            }
+                        Attributes user = new LDAPSearchBuilder(context, domainDN).subTreeScope().searchOne("(& (userPrincipalName={0})(objectCategory=user))", userPrincipalName);
+                        if (user == null) {
+                            // failed to find it. Fall back to sAMAccountName.
+                            // see http://www.nabble.com/Re%3A-Hudson-AD-plug-in-td21428668.html
+                            LOGGER.fine("Failed to find " + userPrincipalName + " in userPrincipalName. Trying sAMAccountName");
+                            user = new LDAPSearchBuilder(context, domainDN).subTreeScope().searchOne("(& (sAMAccountName={0})(objectCategory=user))", samAccountName);
+                            if (user == null) {
+                                throw new UsernameNotFoundException("Authentication was successful but cannot locate the user information for " + username);
+                            }
+                        }
+                        LOGGER.fine("Found user " + username + " : " + user);
 
-            Set<GrantedAuthority> groups = resolveGroups(domainDN, dnFormatted, context);
-            groups.add(SecurityRealm.AUTHENTICATED_AUTHORITY);
+                        Object dnObject = user.get(DN_FORMATTED).get();
+                        if (dnObject == null) {
+                            throw new AuthenticationServiceException("No distinguished name for " + username);
+                        }
 
-            return new ActiveDirectoryUserDetail(username, password, true, true, true, true, groups.toArray(new GrantedAuthority[groups.size()]),
-                    getStringAttribute(user, "displayName"),
-                    getStringAttribute(user, "mail"),
-                    getStringAttribute(user, "telephoneNumber")
-                    ).updateUserInfo();
-        } catch (NamingException e) {
-            if (anonymousBind && e.getMessage().contains("successful bind must be completed") && e.getMessage().contains("000004DC")) {
-                // sometimes (or always?) anonymous bind itself will succeed but the actual query will fail.
-                // see JENKINS-12619. On my AD the error code is DSID-0C0906DC
-                throw new UserMayOrMayNotExistException("Unable to retrieve the user information without bind DN/password configured");
-            }
+                        String dn = dnObject.toString();
+                        LdapName ldapName = new LdapName(dn);
+                        String dnFormatted = ldapName.toString();
+
+                        if (bindName != null && !password.equals(NO_AUTHENTICATION)) {
+                            // if we've used the credential specifically for the bind, we
+                            // need to verify the provided password to do authentication
+                            LOGGER.fine("Attempting to validate password for DN=" + dn);
```

Comment:
Don't perform string concatenation in conditional logging statements.
Replace with `LOGGER.log(Level.FINE, "blah {}", dn);` or whatever the correct idoyim is for parameterized logging in the logging framework that is being used.